genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcAgentJobKillServiceImpl.java

```diff
@@ -17,117 +17,165 @@
  */
 package com.netflix.genie.agent.execution.services.impl.grpc;
 
-import com.google.common.util.concurrent.FutureCallback;
-import com.google.common.util.concurrent.Futures;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.netflix.genie.agent.execution.services.AgentJobKillService;
 import com.netflix.genie.agent.execution.services.KillService;
+import com.netflix.genie.common.internal.util.ExponentialBackOffTrigger;
 import com.netflix.genie.proto.JobKillRegistrationRequest;
 import com.netflix.genie.proto.JobKillRegistrationResponse;
 import com.netflix.genie.proto.JobKillServiceGrpc;
 import lombok.extern.slf4j.Slf4j;
-import org.springframework.core.task.TaskExecutor;
+import org.springframework.scheduling.TaskScheduler;
 
 import javax.validation.constraints.NotBlank;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 /**
- * Implementation of the {@link AgentJobKillService}.
+ * Implementation of the {@link AgentJobKillService}, listens for kill coming from server using long-polling.
  *
- * @author standon
+ * Note: this implementation still suffers from a serious flaw: because it is implemented with a unary call,
```

Comment:
Is there something stopping us from doing this now rather than putting something in we think won't really be right?

Reply:
Just doing cost/benefit, there are more urgent things to work on.
A small accumulation of parked requests that take a long time to timeout on the server hasn't created problems.
Fixing this would require a protocol change which along with migration of existing clients makes the cost of fixing this pretty high.